### PR TITLE
[GPU Process] [FormControls] Add a ControlPart for SearchFieldResultsButton and SearchFieldResultsDecoration

### PR DESCRIPTION
**Pull request by shallawa**
#### bb20f68f21269357d577a49f976ab8ed8dd7c99c
<pre>
[GPU Process] [FormControls] Add a ControlPart for SearchFieldResultsButton and SearchFieldResultsDecoration
<a href="https://bugs.webkit.org/show_bug.cgi?id=251373">https://bugs.webkit.org/show_bug.cgi?id=251373</a>
rdar://104825004

Reviewed by Aditya Keerthi.

This ControlPart will handle drawing the SearchField results button and the
SearchField results decoration which will be created for the following markup:

&apos;&lt;input type=&quot;search&quot; results=&quot;0&quot;&gt;&apos;
&apos;&lt;input type=&quot;search&quot; results=&quot;num&quot;&gt;&apos; where num &gt; 0 respectivly.

The existing position calculations are not needed and seem to be incorrect.

For macOS, AppKit will be used to draw the platform controls.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/controls/ControlFactory.h:
* Source/WebCore/platform/graphics/controls/SearchFieldResultsPart.h: Added.
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
(WebCore::ControlFactoryMac::searchMenuTemplate const):
(WebCore::ControlFactoryMac::createPlatformSearchFieldResults):
* Source/WebCore/platform/graphics/mac/controls/SearchFieldResultsMac.h: Added.
* Source/WebCore/platform/graphics/mac/controls/SearchFieldResultsMac.mm: Added.
(WebCore::SearchFieldResultsMac::SearchFieldResultsMac):
(WebCore::SearchFieldResultsMac::updateCellStates):
(WebCore::SearchFieldResultsMac::draw):
* Source/WebCore/platform/mac/ThemeMac.h:
* Source/WebCore/platform/mac/ThemeMac.mm:
(WebCore::drawCellFocusRingWithFrameAtTime): Deleted.
(WebCore::drawCellFocusRing): Deleted.
(WebCore::drawCellOrFocusRingIntoRectWithView): Deleted.
(WebCore::ThemeMac::drawCellOrFocusRingWithViewIntoContext): Deleted.
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::createControlPart const):
* Source/WebCore/rendering/RenderThemeMac.h:
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::canPaint const):
(WebCore::RenderThemeMac::canCreateControlPartForRenderer const):
(WebCore::convertToPaintingPosition): Deleted.
(WebCore::RenderThemeMac::paintCellAndSetFocusedElementNeedsRepaintIfNecessary): Deleted.
(WebCore::RenderThemeMac::setSearchCellState): Deleted.
(WebCore::RenderThemeMac::paintSearchFieldDecorationPart): Deleted.
(WebCore::RenderThemeMac::paintSearchFieldResultsDecorationPart): Deleted.
(WebCore::RenderThemeMac::paintSearchFieldResultsButton): Deleted.
(WebCore::RenderThemeMac::search const): Deleted.
(WebCore::RenderThemeMac::searchMenuTemplate const): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;ControlPart&gt;::decode):

Canonical link: <a href="https://commits.webkit.org/259779@main">https://commits.webkit.org/259779@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/aa3f11ffe09419d206c1b440849fc76e2931c749

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105905 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14973 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38751 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115099 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/175220 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109807 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16391 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/6166 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98133 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114854 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111655 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95449 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94338 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27093 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/81693 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8237 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/28447 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8719 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/6166 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14349 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47987 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6767 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10273 "Built successfully") | | | 
| | | | |

### CODE REVIEW

File: Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h [hunk 1]
```diff
@@ -35,6 +35,7 @@ OBJC_CLASS NSServicesRolloverButtonCell;
 namespace WebCore {
 
 class ControlFactoryMac final : public ControlFactoryCocoa {
+    friend class SearchFieldResultsMac;
```
Comment (pxlcoder) [inline Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h]: Why is this necessary?

**Review by pxlcoder — COMMENTED** [review]
Did you test RTL?

**Comment by shallawa** [comment]
> Did you test RTL?

Yes. With this patch, the test case https://bug-251373-attachments.webkit.org/attachment.cgi?id=464755 works as expected. When typing text in the search field, the text is displayed nicely between the results button and the cancel button.

File: Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h (hunk 1 above)
Comment (shallawa) [reply Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h]: To make `SearchFieldResultsMac` have access to the new private method `ControlFactoryMac::searchMenuTemplate()`. I did not want to make it public since it is only used by `SearchFieldResultsMac::updateCellStates()`.

Comment (pxlcoder) [reply Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h]: I see. Can `searchMenuTemplate()` be a private method on `SearchFieldResultsMac`?

I generally prefer to avoid use of `friend` when possible.

Comment (pxlcoder) [reply Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h]: Or alternately, pass it in to the constructor?

**Comment by pxlcoder** [comment]
Thanks for the screenshots!

**Review by pxlcoder — APPROVED** [review]

File: Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h (hunk 1 above)
Comment (shallawa) [reply Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h]: `searchMenuTemplate` is now passed to `SearchFieldResultsMac` constructor as an argument.
